Удаление просроченных рекламных постов BJS2 41309 Ad

SerUKalinin opened the pull request.

SerUKalinin committed:
src/main/java/faang/school/postservice/PostServiceApp.java

SerUKalinin committed:
src/main/java/faang/school/postservice/config/async/AsyncConfig.java

SerUKalinin committed:
src/main/java/faang/school/postservice/config/async/AsyncConfig.java

SerUKalinin committed:
src/main/java/faang/school/postservice/repository/ad/AdRepository.java

SerUKalinin committed:
src/main/java/faang/school/postservice/scheduler/ScheduledExpiredAdRemover.java

SerUKalinin committed:
src/main/java/faang/school/postservice/service/ad/AdService.java

SerUKalinin committed:
src/main/java/faang/school/postservice/service/ad/AdServiceImpl.java

SerUKalinin committed:
src/main/resources/application.yaml

SerUKalinin committed:
src/test/java/faang/school/postservice/scheduler/ScheduledExpiredAdRemoverTest.java

SerUKalinin committed:
src/test/java/faang/school/postservice/service/ad/AdServiceImplTest.java

SerUKalinin committed:
Merge branch 'Titan-master-stream7' into BJS2-41309-Ad

DaurenBerdibekov commented on src/main/java/faang/school/postservice/scheduler/ScheduledExpiredAdRemover.java:
```diff
@@ -0,0 +1,39 @@
+    private final AdRepository adRepository;
+
+    @Scheduled(cron = "${post.ad-remover.scheduler.cron}")
+
```
Думаю тут надо эту пустую строку убрать

DaurenBerdibekov commented on src/main/java/faang/school/postservice/scheduler/ScheduledExpiredAdRemover.java:
```diff
@@ -0,0 +1,39 @@
+        List<Long> adsToDelete = adRepository.findAllExpiredIds();
+
+        if (adsToDelete.isEmpty()) {
+            return;
```
тут надо бы чтото возвращать, скорее всего исключение. И можно еще поставить лог

DaurenBerdibekov commented on src/main/java/faang/school/postservice/scheduler/ScheduledExpiredAdRemover.java:
```diff
@@ -0,0 +1,39 @@
+    @Scheduled(cron = "${post.ad-remover.scheduler.cron}")
+
+    public void deleteExpiredAdsScheduled() {
+        List<Long> adsToDelete = adRepository.findAllExpiredIds();
```
Хоть в задании и сказано что надо доставать все, я бы наверное тут все таки не доставал прям все, а переписал запрос и сразу бы доставал рекламу которую надо удалить.

DaurenBerdibekov commented on src/main/java/faang/school/postservice/service/ad/AdServiceImpl.java:
```diff
@@ -0,0 +1,25 @@
+    @Transactional(propagation = Propagation.REQUIRES_NEW)
+    public void deleteAds(List<Long> adIds) {
+        adRepository.deleteAllById(adIds);
+    }
```
я бы добавил обработку исключений также сюда и проверку входных данных

DaurenBerdibekov reviewed (approved):
На мой взгляд все четко)

SerUKalinin committed:
удалил лишнюю строку в ScheduledExpiredAdRemover

SerUKalinin committed:
List<Ad> вместо List<Long> findAllExpiredAds

SerUKalinin committed:
@Query("SELECT a FROM Ad a WHERE a.endDate <CURRENT_TIMESTAMP OR a.appearancesLeft <=0") вместо     @Query("SELECT a.id FROM Ad a WHERE a.endDate <CURRENT_TIMESTAMP OR a.appearancesLeft <=0")

SerUKalinin committed:
1) List<Ad> adsToDelete = adRepository.findAllExpiredAds();

вместо

List<Long> adsToDelete = adRepository.findAllExpiredIds();

2) List<List<Ad>> partitionedAdsToDelete = ListUtils.partition(adsToDelete, maxListSize);

вместо

List<List<Long>> partitionedAdsToDelete = ListUtils.partition(adsToDelete, maxListSize);

SerUKalinin committed:
void deleteAds(List<Ad> ads);

вместо

void deleteAds(List<Long> adIds);

SerUKalinin committed:
Merge remote-tracking branch 'origin/BJS2-41309-Ad' into BJS2-41309-Ad

SerUKalinin committed:
public void deleteAds(List<Ad> ads) {
adRepository.deleteAll(ads);
}

вместо

public void deleteAds(List<Long> adIds) {
adRepository.deleteAllById(adIds);
}

SerUKalinin committed:
Доработал тесты

SerUKalinin committed:
Доработал тест

SerUKalinin committed:
Добавил логирование и исключения

SerUKalinin committed:
Добавил @RequiredArgsConstructor

SerUKalinin committed:
Удалил @RequiredArgsConstructor

SerUKalinin committed:
Удалил @RequiredArgsConstructor

SerUKalinin committed:
Добавил логирование и исключения

SerUKalinin committed:
Доработал тест

zaqqaz226 commented on src/main/java/faang/school/postservice/config/async/AsyncConfig.java:
```diff
@@ -0,0 +1,16 @@
+@Configuration
+public class AsyncConfig {
+    @Bean
+    public ExecutorService adRemoverExecutorService(@Value("${post.ad-remover.threads-count}") int threadsCount) {
```
Вынеси в отдельную переменную - @Value("${post.ad-remover.threads-count}") int threadsCount. Это не ошибка, но смотреться будет гораздо лучше.

SerUKalinin replied:
Готово, даже не подумал чет об этом. Спасибо

zaqqaz226 commented on src/test/java/faang/school/postservice/scheduler/ScheduledExpiredAdRemoverTest.java:
```diff
@@ -0,0 +1,81 @@
+    void deleteExpiredAdsScheduledException() {
+        when(adRepository.findAllExpiredAds()).thenThrow(new RuntimeException("Database error"));
+
+        try {
```
А чего не через assertThrows?

SerUKalinin replied:
буду знать что это более правильное решение, почитал про него понял разницу)

SerUKalinin committed:
Вынес в отдельную переменную @Value("${post.ad-remover.threads-count}") int threadsCoun

SerUKalinin committed:
заменил использование try-catch на assertThrows

zaqqaz226 commented on src/main/java/faang/school/postservice/config/async/AsyncConfig.java:
```diff
@@ -0,0 +1,18 @@
+public class AsyncConfig {
+    @Value("${post.ad-remover.threads-count}")
+    private int threadsCount;
+    @Bean
```
Поставь тут энетер менжду строками пж

SerUKalinin committed:
добавил строку
